### PR TITLE
feat(vx-scan): add Custom scanner support

### PR DESCRIPTION
**Pull request by eventualbuddha**
## Overview

Adds `custom-scanner` library and uses it from a new `vx-scan` state machine. Keeps support for PlusTek using `SCANNER_MODEL=plustek`. Best to review one commit at a time.

## Demo Video or Screenshot
https://user-images.githubusercontent.com/1938/216479861-6f0f67ac-c0c7-4d71-ab15-0549658d6642.MOV

## Testing Plan
Manual testing:
- [x] happy path: properly marked ballot fed correctly is accepted
- [x] overvote is rejected
- [ ] non-ballot sheet is rejected
- [ ] all orientations interpreted

Automated testing:
- [ ] add tests for `custom-scanner`
- [ ] add tests for Custom `vx-scan` state machine

## Checklist
- [ ] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.
- [ ] I have added JSDoc comments to any newly introduced exports

- [ ] I have added a screenshot and/or video to this PR to demo the change
- [ ] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

**Comment by eventualbuddha** [comment]
@adghayes this is not yet ready to be merged, but it's getting pretty close to the likely final API. Mind taking a look to see if it makes sense to you? `openScanner` is the main entry point for the library. It returns a `CustomScanner` that has high-level methods like `scan`, `getStatus`, and `move`. That sits on top of a `CustomDevice` which is a simplified abstraction providing i.e. `read` and `write` methods, and itself sits on top of a `WebUSBDevice` that handles the low-level USB. I only put in a little bit of effort to make this extensible beyond the A4 scanner, mostly by trying not to box us in too much.

The library exports its own mocks for each of the levels mentioned earlier, but they're not used in the scan backend yet.

File: libs/custom-scanner/src/custom_scanner.ts [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+          scanParameters
+        );
+      }
+
+      if (setScanParametersResult.isErr()) {
+        return setScanParametersResult;
+      }
+    }
+
+    const { unlock } = await this.publicApiMutex.asyncLock();
```
Comment (adghayes) [inline libs/custom-scanner/src/custom_scanner.ts:451]: Why do this here are not at the top? Just curious.

**Review by adghayes — COMMENTED** [review]
@eventualbuddha the levels of abstraction with the scanner make a lot of sense to me. 

It took me a moment to understand what you're doing with separate public and private locks. I see that it makes the `CustomScanner` the authority on anything that goes wrong during the scan - it's not the consumer's job to poll the scanner to see if anything has gone wrong during the scan. Makes sense.

I also appreciate that this is just a lib and not a separate process.

Is image data returned from the scanner one side at a time? I see a lot of `ScanSide` logic but, in our our use case, we're almost always going to be using `ScanSide.A_and_B` right?

Anyway, LGTM.

File: libs/custom-scanner/src/custom_device.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+      return writeResult;
+    }
+
+    return this.read(maxLength);
+  }
+
+  /**
+   * Gets the read and write endpoints for communicating with the device.
+   */
+  private getReadWriteEndpoints(): Result<
```
[the remark this answers is not in the document]
Comment (eventualbuddha) [reply libs/custom-scanner/src/custom_device.ts]: This is a fair criticism. What it's really doing is validating the endpoints. I changed it in bbbd8e68d604e4ca0f1d668a782166068690a874 to make it clearer.

File: libs/custom-scanner/src/open_scanner.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    );
+    const connectResult = await customA4ScannerDevice.connect();
+
+    if (connectResult.isErr()) {
+      debug('connection error: %o', connectResult.err());
+      return connectResult;
+    }
+
+    debug('connected to device');
+    return ok(new CustomScanner(customA4ScannerDevice));
```
[the remark this answers is not in the document]
Comment (eventualbuddha) [reply libs/custom-scanner/src/open_scanner.ts:38]: I agree. I found it a little confusing even when I was deep in writing all this, so there's definitely some room for improvement. Like you, I'm not sure what better names would be. Do @jonahkagan or @benadida want to drive-by suggest something?

File: libs/custom-scanner/src/custom_scanner.ts (hunk 1 above)
Comment (eventualbuddha) [reply libs/custom-scanner/src/custom_scanner.ts:451]: No good reason. I'll move it.

File: libs/custom-scanner/src/open_scanner.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (benadida) [reply libs/custom-scanner/src/open_scanner.ts:38]: I would lightly suggest `Scanner`, as `Device` is indeed too generic. I don't think we're going to confuse `Scanner` with `VxScan` in our code. So `PlustekScanner` and `CustomScanner` in this case, and in other cases move away from `Device` and to names that are more specific. We can use `Device` when we actually mean to be generic, e.g. a USB device that could be a scanner, printer, or USB stick, that's great. But if it's a disk, I'd say `UsbDisk`.

[the remark this answers is not in the document]
Comment (eventualbuddha) [reply libs/custom-scanner/src/open_scanner.ts:57]: Okay, I've renamed our "device" to "channel" as in "IO channel". That way the stack is:
- scanner
- channel
- device

**Comment by eventualbuddha** [comment]
This branch is no longer targeting `main` and so we do not expect all tests to be passing.

**Review by carolinemodic — APPROVED** [review]

**Pull request merged**